handle multi-valued headers correctly

wfurt opened the pull request:
fixes #34377
related to #35164 

write out multi-valued headers as single header.

new tests waiting for GenericLoopback to land.

stephentoub commented:
@dotnet-bot test this please

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -20,6 +23,33 @@ internal static class HPackEncoder
+        }
+
+        public static bool EncodeHeader(string name, string[] values, string separator, Span<byte> buffer, out int length)
+        {
```
Debug.Assert(values != null), to match the similar check for value in the other overload?

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -48,21 +78,36 @@ public static bool EncodeHeader(string name, string value, Span<byte> buffer, ou
-        private static bool EncodeString(string s, Span<byte> buffer, out int length, bool lowercase)
+        private static bool EncodeStringPart(string s, Span<byte> buffer, ref int i, bool lowercase)
         {
             const int toLowerMask = 0x20;
```
Nit: toLowerMask => ToLowerMask

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -48,21 +78,36 @@ public static bool EncodeHeader(string name, string value, Span<byte> buffer, ou
 
+                //buffer[i++] = (byte)(s[j] | (lowercase && s[j] >= 'A' && s[j] <= 'Z' ? toLowerMask : 0));
+                buffer[i++] = (byte)(s[j] | (lowercase ? toLowerMask : 0));
+            }
```
Let's add a TODO to use https://github.com/dotnet/corefx/issues/34144 when it's available.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -48,21 +78,36 @@ public static bool EncodeHeader(string name, string value, Span<byte> buffer, ou
+                if (i >= buffer.Length)
+                {
+                    return false;
+                }
```
It'd be better to do this check once outside of the loop, e.g. create a span from the string, check whether that span is longer than buffer.Length - i, and if it is, slice the span, then iterate over the span.  That way we don't do this check as part of each iteration.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -48,21 +78,36 @@ public static bool EncodeHeader(string name, string value, Span<byte> buffer, ou
+                }
 
+                //buffer[i++] = (byte)(s[j] | (lowercase && s[j] >= 'A' && s[j] <= 'Z' ? toLowerMask : 0));
+                buffer[i++] = (byte)(s[j] | (lowercase ? toLowerMask : 0));
```
It'd probably be better to have two loops, one if lowercase is true and one if it's false.

geoffkizer replied:
Going further, I would suggest having two entirely different routines here. Or possibly three.

There are really three cases here:
(1) Encode header name. Always a single value. Checked for validity on construction, so we don't need to check again here (though an assert would be good). Needs to be lower cased, but current lower casing logic is incorrect. See #35164. Note also, ideally this is not the common case because we should optimize known header values, see #31307.
(2) Encode single header value. Always a single value. No lower casing. Needs to be checked for valid ASCII, as we do in HttpConnection but aren't doing here today. See #35165.
(3) Encode multiple header values. Again, needs to be checked for valid ASCII.

Possibly (2) and (3) are the same routine.

wfurt replied:
I'll take a look. I generally preserved existing code but I saw #3516 so I added extra check.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -48,21 +78,36 @@ public static bool EncodeHeader(string name, string value, Span<byte> buffer, ou
+                }
 
+                //buffer[i++] = (byte)(s[j] | (lowercase && s[j] >= 'A' && s[j] <= 'Z' ? toLowerMask : 0));
+                buffer[i++] = (byte)(s[j] | (lowercase ? toLowerMask : 0));
```
`i` here is a ref; it will probably be faster to store it into a local, iterate on the local, and then store the local back to `i` when you're done.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -73,22 +118,47 @@ private static bool EncodeString(string s, Span<byte> buffer, out int length, bo
+            }
+            else
+            {
+                // Calculate length of all pars and separators.
```
Nit: pars => parts

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -73,22 +118,47 @@ private static bool EncodeString(string s, Span<byte> buffer, out int length, bo
+                    i += part.Length;
+                }
+                i += (parts.Length - 1) * separator.Length;
+                s = parts[0];
```
Is parts guaranteed not to be empty?

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -699,6 +700,32 @@ private void WriteHeader(string name, string value)
             _headerBuffer.Commit(bytesWritten);
         }
 
+        private void WriteHeader(string name, string[] value, string separator)
```
Is this actually used? Where?

geoffkizer commented:
We should enable the tests marked with #34377 and ensure they pass with this change.

wfurt commented:
I added #35266 to track missing tests for this and previous PR. I can work on #35165 and #31307 and add tests as part of that work. I did not want to inflate this change.

wfurt commented:
I updated SendAsync_UserAgent_CorrectlyWritten to use GenericLoopback server and it can run now properly:

```
 System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_HttpClientHandlerTest_Http2.SendAsync_UserAgent_CorrectlyWritten [STARTING]
      System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_HttpClientHandlerTest_Http2.SendAsync_UserAgent_CorrectlyWritten [FINISHED] Time: 0.0139004s
```

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
             length = 0;
 
-            if (buffer.Length == 0)
+            // We need at least \0 and twice length plus one octet string
```
Is this actually true? I think that having an empty (zero-length) value is allowed.

wfurt replied:
I was looking for the answer and it seems like older spec did allow it but current specification does not. 
In worst case we will ask for buffer than needed if this really fits on buffer size boundary, right?

geoffkizer replied:
Where does the RFC disallow this?

It doesn't really matter either way, since we will check again below. But we should understand what's allowed and what's not.

wfurt replied:
I found discussion about on Internet. Old HTTP/1.1 allowed empty and NULL values. I also tested zero length value headers again nginx and Azure IIS servers and they both happily take them. 

I updated code accordingly.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
             }
 
-            if (!EncodeString(value, buffer.Slice(i), out int valueLength, lowercase: false))
+            if (currentIndex + s.Length >= buffer.Length)
```
Shouldn't this be `>` instead of `>=`?

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
+        private static bool EncodeStringPart(string s, Span<byte> buffer, ref int currentIndex)
         {
-            const int toLowerMask = 0x20;
+            if (s.Length >= buffer.Length)
```
Same question as above

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
-            buffer[0] = 0;
+            if (s != null)
+            {
+                i = s.Length;
```
Instead of `i` can we call this something more descriptive, like `headerValueLength` or similar?

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
+            return true;
+        }
+
+        private static bool EncodeHeaderValue(string s, string[] parts, string separator, Span<byte> buffer, out int length)
```
It seems like we should have two different routines here, one for the single-value case and one for the multi-value case. They are different enough that this seems worthwhile, and that way we son't have multiple conditionals inside the routine.

wfurt replied:
I did split the functions. It seems like there will be some duplication but each seems cleaner now.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
 
-        private static bool EncodeString(string s, Span<byte> buffer, out int length, bool lowercase)
+        // Encode fragment or header value without writing out length.
+        private static bool EncodeStringPart(string s, Span<byte> buffer, ref int currentIndex)
```
The pattern for these routines is not to pass `ref int currentIndex`, but instead to pass the Span starting at the appropriate place for the next byte to write, and return the length via `out int length`. 

We should be consistent about the pattern. If there's a good reason to change the pattern (e.g. perf) then we should change it everywhere.

BTW, it seems like there's a bug directly below related to this: the length check does not consider `currentIndex`.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
+            return EncodeHeader(name, value, null, null, buffer, out length);
+        }
+
+        public static bool EncodeHeader(Headers.HeaderDescriptor header, string value, Span<byte> buffer, out int length)
```
It looks like these overloads (here and directly below) that take HeaderDescriptor are not used. Can we remove them?

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -701,6 +701,19 @@ private void WriteHeader(string name, string value)
             _headerBuffer.Commit(bytesWritten);
         }
 
+        private void WriteHeader(HeaderDescriptor header, string[] value, string separator)
```
For consistency with the above routine, seems like this should take `string name` instead of `HeaderDescriptor header`.

geoffkizer replied:
Also `value` => `values`

wfurt commented:
I did another update to
- react to feedback from review
- filter out more header banned by RFC
-  add minimal support to GenericLoopback for POST
- fixed and extended header test.
- added test case for #35164 (so this fixes #35164 as well)

geoffkizer commented on src/Common/tests/System/Net/Http/GenericLoopbackServer.cs:
```diff
@@ -85,5 +89,35 @@ public int GetHeaderValueCount(string headerName)
+            if (buffer == null)
+            {
+                // In most cases we should get body in single frame for tests.
+                _body = ArrayPool<byte>.Shared.Rent(data.Length);
```
There's no need to rent buffers in tests. Just allocate a byte[].

geoffkizer commented on src/Common/tests/System/Net/Http/Http2LoopbackServer.cs:
```diff
@@ -515,6 +520,15 @@ public async Task<(int streamId, HttpRequestData requestData)> ReadAndParseReque
+            while ((frame.Flags & FrameFlags.EndStream) == 0)
+            {
+                frame = await ReadFrameAsync(Timeout).ConfigureAwait(false);
+                if (frame.Type == FrameType.Data && frame.Length > 1)
```
This is going to discard any non-DATA frame. That seems bad.

wfurt replied:
should we throw for now?

geoffkizer commented on src/Common/tests/System/Net/Http/GenericLoopbackServer.cs:
```diff
@@ -85,5 +89,35 @@ public int GetHeaderValueCount(string headerName)
+
+        public Span<byte> Body => _body.AsSpan().Slice(0, _bodyLength);
+
+        internal void AddBodyData(ReadOnlySpan<Byte> data)
```
I don't think we should add this logic here. HttpRequestData is intentionally not very smart. Just expose something like `public Memory<byte> Body>` similar to method and path above. This logic for constructing the body should live in Http2LoopbacKServer.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -454,32 +463,80 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
             {
-                throw new NotImplementedException("Request body not supported");
-            }
+                headerLines = await connection.ReadRequestHeaderAndSendResponseAsync(statusCode, headerString + "Connection: close\r\n", content);
```
It seems weird to send the response before you read the request body. This may work ok for some tests, but in general it seems like we should do the normal thing here and read the body before sending the response.

wfurt replied:
I agree. It was convenience to keep it as it was. I split the call and we send response now at the end.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -454,32 +463,80 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
+
+                        while (contentLength != 0)
+                        {
+                            int length = await connection.Reader.BaseStream.ReadAsync(buffer, offset, contentLength);
```
We usually call this `bytesRead`.

Also, it might be worthwhile to assert that the bytesRead doesn't exceed contentLength, otherwise you will probably end up hanging here.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -454,32 +463,80 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
+                            string chunkHeader = await connection.Reader.ReadLineAsync();
+                            if (chunkHeader == null)
+                            {
+                                // Some tests do not send complete body.
```
Which tests don't send a complete body? It seems bad to just always allow this, because it might hide bugs.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -454,32 +463,80 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
+                            int contentLength = int.Parse(chunkHeader, System.Globalization.NumberStyles.HexNumber);
+                            if (contentLength == 0)
+                            {
+                                // Last chunk
```
Shouldn't we read the trailing \r\n here, so we ensure we validate the complete request body?

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -454,32 +463,80 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
+                                // Last chunk
+                                break;
+                            }
+                            char[] buffer = new char[contentLength];
```
Why is this char instead of byte?

wfurt replied:
I did not find good method in StreamReader. When I used StreamReader.BaseStream I was getting incomplete data. It seems like the StreamReader can consume some and store them in internal buffer. 
If you or @stephentoub  have any specific suggestion, please let me know.

wfurt replied:
It turned out is was bad idea to mic StreamReader.ReadAsync/ReadLine with StreamReader.BaseStream.ReadAsync. It would be nice to get byte[] back using IO on underlying stream but so far I did not find reliable way how to make it happen.
So for now I always use  ReadAsync returning char[] and converting to byte[].

geoffkizer replied:
We really shouldn't be using StreamReader to implement the loopback server. I forgot we were doing this. We should change this so we don't use StreamReader, but that's a fair amount of work to do, unfortunately.

What's driving you to add support for reading the request body here? It would be nice to have this, but I'm not sure why it's needed as part of this PR specifically. And if we are going to add this support, it seems like we should do it properly, rather than hack it.

wfurt replied:
The test I was trying to fix and use was using POST. 
It does not have to but I thought it would be nice to get it in so we can keep progressing on fixing tests. 
In this particular case I can change it to GET or disable this for HTTP/1.x 

Let me know what direction do you want to take @geoffkizer

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -454,32 +463,80 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
+                            }
+                            char[] buffer = new char[contentLength];
+                            int offset = 0;
+                            while (contentLength != 0)
```
You could probably factor this logic out so it could be shared between the Content-Length case and the chunked case.

wfurt replied:
I simplified this with ReadBlockAsync so there is no longer need to loop.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -454,32 +463,80 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
+                    }
+                    else
+                    {
+                        await connection.Drain();
```
Is this needed? Seems like this should be an error.

wfurt replied:
This was attempt to mimic actual handler. I removed it for now. We can revisit this later as we convert more tests.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -24,61 +24,110 @@ internal static class HPackEncoder
-        }
+            int i = 0;
+            length = 0;
+            int valueLength = (value == null ? 0 : value.Length);
```
Can value actually be null here, or should it be an empty string instead?

wfurt replied:
I was not sure. But since we have test now and it does get empty value I updated this line.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -24,61 +24,110 @@ internal static class HPackEncoder
 
-            // We need at least \0 and twice length plus one octet string
-            if (buffer.Length < 5)
+            if (values == null || values.Length == 0)
```
Can values actually be null here?

wfurt replied:
I was not sure but since we also check for empty array it seems like it would be cheap and safe to keep it in.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -24,61 +24,110 @@ internal static class HPackEncoder
-            return EncodeHeader(header.Name, value, null, null, buffer, out length);
-        }
+            // We need at least \0 and twice length plus one octet string
+            if (buffer.Length < 3 + name.Length + (value == null ? 0 : value.Length))
```
Use `valueLength` here

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -24,61 +24,110 @@ internal static class HPackEncoder
+
+            if (valueLength == 0)
+            {
+                // We are done of there is no value.
```
of -> if

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -128,68 +178,8 @@ private static bool EncodeStringPart(string s, Span<byte> buffer, ref int curren
-            {
-                return false;
-            }
+            encodedLength = i;
```
It's probably better to use a for loop here instead of foreach.

wfurt replied:
is there performance benefit or is it mostly style?

geoffkizer replied:
I think there's a perf benefit, because we don't need to allocate the IEnumerator. But that may be optimized away, not sure.

wfurt replied:
ok. that would make sense.

stephentoub replied:
> But that may be optimized away, not sure.

The C# compiler recognizes foreach loops over arrays, strings, and spans, and turns them into the equivalent of a for loop.  Even if it didn't, we do expose a struct-returning GetEnumerator on span to avoid that allocation, though per the previous sentence it'll rarely be used.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -701,6 +701,19 @@ private void WriteHeader(string name, string value)
 
+        private void WriteHeader(string name, string[] values, string separator)
+        {
+            // TODO: ISSUE 31307: Use static table for known headers
```
Sorry @wfurt, my change that implemented this will end up conflicting with a bunch of your changes, but hopefully it'll be fairly straightforward for you to rebase and adapt.

wfurt replied:
updated @stephentoub.  It seems like the biggest blocking issue is debate around StreamReader and underlying Loopback server.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -223,6 +213,113 @@ public static bool EncodeStringLiteral(string value, Span<byte> destination, out
             return false;
         }
 
+        private static bool EncodeStringLiteralStringPart(string value, Span<byte> destination, out int bytesWritten)
```
What does "StringPart" mean?  Can that just be "Value"?

wfurt replied:
updated.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -223,6 +213,113 @@ public static bool EncodeStringLiteral(string value, Span<byte> destination, out
+
+                if (values.Length == 0)
+                {
+                    return EncodeStringLiteral(null, destination, out bytesWritten);
```
null?  Won't that result in a null reference exception from https://github.com/dotnet/corefx/pull/35202/files#diff-ddec2c87b42f404061f4cd18008cfcc7R253?

wfurt replied:
It seems like when I pass null to Add, it does get converted to "". So I updated code to do the same.

wfurt commented:
outerloop tests ok:
```
  === TEST EXECUTION SUMMARY ===
     System.Net.Http.Functional.Tests  Total: 6902, Errors: 0, Failed: 0, Skipped: 38, Time: 2207.632s
  ~/git/wfurt-corefx2/src/System.Net.Http/tests/FunctionalTests
```

geoffkizer commented on src/Common/tests/System/Net/Http/Http2LoopbackServer.cs:
```diff
@@ -515,6 +520,33 @@ public async Task<(int streamId, HttpRequestData requestData)> ReadAndParseReque
             requestData.Method = requestData.GetSingleHeaderValue(":method");
             requestData.Path = requestData.GetSingleHeaderValue(":path");
 
+            byte[]  body = null;
```
Nit: spacing

geoffkizer commented on src/Common/tests/System/Net/Http/Http2LoopbackServer.cs:
```diff
@@ -515,6 +520,33 @@ public async Task<(int streamId, HttpRequestData requestData)> ReadAndParseReque
+            while ((frame.Flags & FrameFlags.EndStream) == 0)
+            {
+                frame = await ReadFrameAsync(Timeout).ConfigureAwait(false);
+                // TODO. support more types.
```
What types do you have in mind here?

geoffkizer commented on src/Common/tests/System/Net/Http/Http2LoopbackServer.cs:
```diff
@@ -515,6 +520,33 @@ public async Task<(int streamId, HttpRequestData requestData)> ReadAndParseReque
+                frame = await ReadFrameAsync(Timeout).ConfigureAwait(false);
+                // TODO. support more types.
+                Assert.Equal(FrameType.Data, frame.Type);
+                if (frame.Type == FrameType.Data && frame.Length > 1)
```
Since we assert just above, there's no need to test the frame type again

wfurt replied:
Could we get eventually WINDOW_UPDATE, RST_STREAM or GOAWAY? We probably won't right now but I thought other types may be possible.

geoffkizer replied:
The assumption in this code is that we're expecting just the body of this particular request. As such we shouldn't be receiving anything else here.

geoffkizer commented on src/Common/tests/System/Net/Http/Http2LoopbackServer.cs:
```diff
@@ -515,6 +520,33 @@ public async Task<(int streamId, HttpRequestData requestData)> ReadAndParseReque
+                {
+                    if (body == null)
+                    {
+                        body = ((DataFrame)frame).Data.ToArray();
```
For code clarity, it seems better to cast the frame to a DataFrame once instead of doing it every time you use it

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
 
         public sealed class Connection : IDisposable
         {
+            private const int _blockSize=4000;
```
Nit: spacing

geoffkizer replied:
Also, consts are usually capitalized without _, like so: BlockSize. And maybe call this BufferSize?

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
             private Stream _stream;
-            private StreamReader _reader;
             private StreamWriter _writer;
+            private byte[] _buffer;
```
For clarity, seems like _readBuffer would be better. Also maybe _readStart, _readEnd

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
-            private StreamReader _reader;
             private StreamWriter _writer;
+            private byte[] _buffer;
+            private int _end;
```
Nit: define _start before _end.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
                 _writer = new StreamWriter(stream, Encoding.ASCII) { AutoFlush = true };
+
+                _buffer = new byte[_blockSize];
+                _end= 0;
```
Nit: spacing

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
-            public StreamReader Reader => _reader;
             public StreamWriter Writer => _writer;
 
+            public async Task<int> ReadAsync(byte[] buffer, int offset, int size)
```
Use `Memory<byte>` here

wfurt replied:
updated.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
+                if (_end - _start > 0)
+                {
+                    // Use buffered data first.
+                    int copyLength = _end - _start > size ? size : _end - _start;
```
use Math.Min

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
+                    return copyLength;
+                }
+
+                return await _stream.ReadAsync(buffer,offset, size).ConfigureAwait(false);
```
Nit: spacing

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
+            }
+
+            // Read until we either get requested data or we hit end of stream.
+            public async Task<int> ReadBlockAsync(byte[] buffer, int offset, int size)
```
`Memory<byte>`

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
+
+                    if (readLength == 0)
+                    {
+                        break;
```
Should this be an exception? Seems like this is always an error, right?

wfurt replied:
added exception.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
+                return totalLength;
+            }
+
+            public async Task<int> ReadBlockAsync(char[]  result, int offset, int size)
```
Is this actually used? Looks like you are only using the byte version.

wfurt replied:
same as bellow

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
+                return readLength;
+            }
+
+            public async Task<string> ReadToEndAsync()
```
This seems like a weird method to have. The tests that use this basically want to read the content body. Seems like they should just do that. The fact that they are using ReadToEndAsync currently is really just an artifact of how the loopback server was implemented originally, and we will need to fix this up for HTTP2 anyway.

wfurt replied:
This and  ReadToEndAsync is used directly in some HTTP tests. It is not used by loopback it self.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
-            public StreamReader Reader => _reader;
             public StreamWriter Writer => _writer;
 
+            public async Task<int> ReadAsync(byte[] buffer, int offset, int size)
```
This and a bunch of other methods here are public. Do they need to be? Seems like these should be private implementation details.

wfurt replied:
some methods from what used to be StreamReader are used by individual tests.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
+                        if (_end + 2 > _buffer.Length)
+                        {
+                            // We no longer have space to read CRLF. Allocate new buffer and start over.
+                            byte[] newBuffer = new byte[_buffer.Length + _blockSize];
```
It might simplify things a bit to use ArrayBuffer from the product here.

wfurt replied:
I'll do it as separate change. I assume I should move it to common place to avoid direct linking like we do in UnitTests, right?

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -365,25 +365,167 @@ public class Options
+                    }
+
+                    int stringLength = index - _start;
+                    while (stringLength > 0 && _buffer[_start+stringLength] == '\n' || _buffer[_start+stringLength] == '\r')
```
This seems too loose. Presumably you want to remove the CR, and also the LF if it's present. Can you just do that?

wfurt replied:
updated.

geoffkizer commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -437,6 +578,15 @@ public async Task<List<string>> ReadRequestHeaderAndSendResponseAsync(HttpStatus
                 return lines;
             }
+
+            public async Task Drain(int max = 4000)
```
Is this used?

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cookies.cs:
```diff
@@ -193,26 +178,15 @@ public async Task GetAsync_AddMultipleCookieHeaders_CookiesSent()
 
-        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "Netfx handler does not support custom cookie header")]
+        [ConditionalFact]
```
Why [ConditionalFact] here?

(Actually I'm not sure why you can use [ConditionalFact] without any parameters, what does this even do?)

wfurt replied:
This allows you to use "throw new SkipTestException" to dynamically skip instead of getting "Pass". This is somewhat recent update to MS XUnit extensions.

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cookies.cs:
```diff
@@ -230,7 +204,21 @@ public async Task GetAsync_SetCookieContainerAndCookieHeader_BothCookiesSent()
+                        else
+                        {
+                            // rfc7540 8.1.2.5 states multiple cookie headers should be represented as single value.
+                            cookieHeaderValue = String.Concat(cookieHeaderValue, "; ", header);
```
I think this should only happen for HTTP2, right? Seems like we should ensure the correct behavior here for HTTP/1.1 and HTTP2 independently. I.e. make sure HTTP/1.1 is sending a single header, and HTTP2 is sending multiple (two? or whatever).

wfurt replied:
common function was added.

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cookies.cs:
```diff
@@ -279,9 +256,23 @@ public async Task GetAsync_SetCookieContainerAndMultipleCookieHeaders_BothCookie
+                    string cookieHeaderValue = null;
 
-                    // Multiple Cookie header values are treated as any other header values and are 
+                    foreach (string header in cookieHeaderValues)
```
Since this logic is duplicated, it's probably worth separating out into a separate function -- especially once you add the additional logic I suggest above

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -276,9 +276,10 @@ public async Task SendAsync_GetWithValidHostHeader_Success(bool withPort)
         public async Task SendAsync_GetWithInvalidHostHeader_ThrowsException()
         {
-            if (PlatformDetection.IsNetCore && !UseSocketsHttpHandler)
+            if (PlatformDetection.IsNetCore && (!UseSocketsHttpHandler || LoopbackServerFactory.IsHttp2))
```
Is this temporary? If so we should add a TODO and link to the issue for HTTP2 tests, to ensure this gets cleaned up.

If not, why doesn't it apply to HTTP2? It's true that HTTP2 doesn't send the Host header, but we do inspect the Host header and send it as the "authority" pseudo-header. I'm not clear why this wouldn't apply to HTTP2 as well.

wfurt replied:
This test depends on behavior of external server. For HTTP/1 SSL handshake fails when name does not match. For HTTP/2 I verified that we set SNI to "hostheaderthatdoesnotmatch". But in this case, server return 400 response. Maybe this is one of tests not worth of porting (or making common)

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -826,7 +829,7 @@ public async Task PostAsync_ManyDifferentRequestHeaders_SentCorrectly()
                     request.Headers.TE.Add(new TransferCodingWithQualityHeaderValue("deflate"));
                     request.Headers.Trailer.Add("MyTrailer");
-                    request.Headers.TransferEncoding.Add(new TransferCodingHeaderValue("chunked"));
+                    if (!IsCurlHandler) request.Headers.TransferEncoding.Add(new TransferCodingHeaderValue("chunked"));
```
Why this change?

wfurt replied:
I could not get curl stable with HTTP2 and chunked encoding. I could not reproduce it now, so I move it back as it was before.

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -1650,4 +1649,13 @@ public sealed class SocketsHttpHandlerTest_Cookies_Http2 : HttpClientHandlerTest
     }
+
+    [ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.SupportsAlpn))]
+    public sealed class SocketsHttpHandlerTest_HttpClientHandlerTest_Http2 : HttpClientHandlerTest
```
This means that every single test in HttpClientHanderTest is going to run in HTTP2 loopback mode. The problem is that most of these tests are not generic loopback based tests, so when they run in "HTTP2 mode" they are really just running in HTTP/1.1 mode again. That seems bad.

I believe we need to separate the generic loopback tests into a separate class so that we can specifically run just those tests in HTTP2 mode (as well as HTTP/1.1 of course).

wfurt replied:
my assumption is that we would fix them over time and most of the tests could run for both HTTP/1.x and 2.0. Looking at the file HTTP/1.x tests seems minority and it may be easier to move them instead to HTTP/1 specific file.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -223,6 +213,113 @@ public static bool EncodeStringLiteral(string value, Span<byte> destination, out
+                else if (values.Length == 1)
+                {
+                    return EncodeStringLiteral(values[0], destination, out bytesWritten);
+                }
```
Move this if/else up to the very beginning of the method (before the destination.Length check)?  Otherwise we're paying for that check twice.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -223,6 +213,113 @@ public static bool EncodeStringLiteral(string value, Span<byte> destination, out
+                    valueLength += part.Length;
+                }
+
+                valueLength  += (values.Length - 1) * separator.Length;
```
Nit: extra unnecessary space before the +=

stephentoub replied:
Should these calculations be done with `checked` to avoid any issues related to overflow?

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -768,25 +779,48 @@ private void WriteHeaderCollection(HttpHeaders headers)
+                        if (header.Key.KnownHeader == KnownHeaders.TE)
+                        {
+                            // HTTP/2 allows only 'trailers' TE header. rfc7540 8.1.2.2
+                            foreach (string value  in header.Value)
```
Nit: unnecessary extra space before in

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -768,25 +779,48 @@ private void WriteHeaderCollection(HttpHeaders headers)
+                            // HTTP/2 allows only 'trailers' TE header. rfc7540 8.1.2.2
+                            foreach (string value  in header.Value)
+                            {
+                                if (value == "trailers")
```
This is doing a case-sensitive comparison; does it need to be ordinal case insensitive?

wfurt replied:
I could not find definitive answer. This is not header name so general requirement for case-insensitive comparison does not apply. The spec says:

>   The only exception to this is the TE header field, which MAY be
>    present in an HTTP/2 request; when it is, it MUST NOT contain any
>    value other than "trailers".
>  

Since I could not find any reference suggesting otherwise, I assume this is exact match. 
I don't know if using StringComparison.InvariantCulture would be more appropriate.

rmkerr replied:
I know that we've recieved bug reports where users ran into issues with a similar assumption about header value casing. Unless this is a paticularly hot path, I think it's probably worth making the comparision case insensitive.

davidsh replied:
I thought the HTTP/2 protocol always transmitted headers as lowercase.  So, shouldn't we assume they already are?

wfurt replied:
this is header value. (TE: trailers)

karelz commented:
@wfurt is it ready for re-review?

wfurt commented:
yes.

rmkerr commented on src/Common/tests/System/Net/Http/Http2LoopbackServer.cs:
```diff
@@ -401,7 +402,7 @@ private static int EncodeString(string value, Span<byte> headerBlock)
             new HttpHeaderData("if-none-match", ""),
             new HttpHeaderData("if-range", ""),
-            new HttpHeaderData("if-unmodifiedsince", ""),
+            new HttpHeaderData("if-unmodified-since", ""),
```
Good catch.

wfurt replied:
I would not see it. Test start failing after enabled on HTTP/2

rmkerr commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -106,7 +106,7 @@ public static bool EncodeLiteralHeaderFieldWithoutIndexing(int index, Span<byte>
 
         /// <summary>Encodes a "Literal Header Field without Indexing - New Name".</summary>
-        public static bool EncodeLiteralHeaderFieldWithoutIndexingNewName(string name, string value, Span<byte> destination, out int bytesWritten)
+        public static bool EncodeLiteralHeaderFieldWithoutIndexingNewName(string name, string[] values, string separator, Span<byte> destination, out int bytesWritten)
```
Is this new parameter `separator` ever going to take on a value other than `HttpHeaderParser.DefaultSeparator`? If not, it seems unnecessary.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -223,6 +213,114 @@ public static bool EncodeStringLiteral(string value, Span<byte> destination, out
+                        throw new HttpRequestException(SR.net_http_request_invalid_char_encoding);
+                    }
+
+                    destination[i] = (byte)((uint)(c));
```
Why is the cast to uint needed?

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:
```diff
@@ -223,6 +213,114 @@ public static bool EncodeStringLiteral(string value, Span<byte> destination, out
+                    int encodedLength = 0;
+                    for (int j = 0; j < values.Length; j++)
+                    {
+                        if (j != 0 && !EncodeStringLiteralValue(separator, destination.Slice(integerLength), out encodedLength))
```
As a subsequent optimization, we could store the already-encoded separator bytes along with the header/separator information, and then here just blit it out rather than re-encoding it every time.
